Marlin/src/module/stepper.cpp

```diff
@@ -2460,7 +2460,20 @@ hal_timer_t Stepper::block_phase_isr() {
 
         #if ENABLED(LIN_ADVANCE)
           if (la_active) {
-            const uint32_t la_step_rate = la_advance_steps < current_block->max_adv_steps ? current_block->la_advance_rate : 0;
+            #if ENABLED(LA_ZERO_SLOWDOWN)
+              const uint32_t step_events_left_until_ramp_down = accelerate_before - step_events_completed;
+              const uint32_t max_e_accel =  STEP_MULTIPLY(interval, current_block->max_e_acc); 
```

Comment:
is the `interval` constant across the execution of the whole block? if so, this could be computed only at the beginning of each block.

Reply:
Here is where I have big plans for S_CURVE_ACCELERATION in a follow up PR: I'll keep track of the current acceleration of each step event and compute la_rate from it. That will make linear advance exactly match the s curve and finally be perfect.

Reply:
@tombrazier I tried this, and the resulting quality is fantastic.
It's rather simple:

* I removed all the LA code for acceleration/deacceleration/cruise/new_block from stepper.cpp 
* stored current_step_rate in a variable so I can access it at the end regardless of the current phase
* added this at the end of block_phase_isr
 
```cpp
if (la_active && current_block) {
      // K = mm of filament compression needed per 1mm/s extrusion speed [mm/mm/s].
      // K = steps of filament compression needed per 1steps/time extrusion speed [steps/steps/time].
      
      float e_step_rate = float(curr_step_rate) * xy_to_e_steps;
      float dt = float(interval)/float(STEPPER_TIMER_RATE);
      float dt_inv = 1.0f / dt; 

      float target_la_step_count = k * e_step_rate;
      float distance_to_target = target_la_step_count - current_la_step_count;

      float one_shot_v = distance_to_target * dt_inv;
      float a = ABS(one_shot_v - current_la_step_rate) * dt_inv;
      a = a > a_max ? a_max : a;
      
      float stopping_distance = current_la_step_rate*current_la_step_rate / (2 * a_max);
      
      if (ABS(distance_to_target) <= stopping_distance){
          // If we're within stopping distance, decelerate
          a *= current_la_step_rate<0 ? 1 : -1;
      }
      else{
          // Otherwise, accelerate towards the target
          a *= distance_to_target<0 ? -1 : 1;
      }

      current_la_step_rate += a * dt;
      current_la_step_count += current_la_step_rate * dt;

      const int32_t la_step_rate = current_la_step_rate * e_to_xy_steps;
      set_la_interval((int32_t)curr_step_rate + la_step_rate);
    } 
```

The idea is simple: 
* keep track of the current nozzle pressure
* calculate target nozzle pressure
* accelerate/decelerate extruder to achieve target nozzle pressure
* aim at reaching target nozzle pressure with zero velocity (to avoid overshooting) 

This will always lag a bit, but the resulting quality at full speed is already impressive (better than the code in this PR)

What I did is obviously extremely inefficient (i have a 550MHz mcu) but the results prove that there's a lot of quality to win from full speed LA. The quality with and without S_CURVE was pretty much the same

Reply:
I can see that working very nicely for a fast 32 bit processor, especially if it has floating point instructions.

Reply:
I'll make prints and pics tomorrow, I think I'm getting better results than the current (slower) LA, even though this method takes 30% less time to print.
I hope I can figure out how to make this in the planner so this can be efficient. One thing I believe to be making a difference too is that it corrects for jerk, so it reduces blobs a  bit. It will need to look at contiguous blocks.

Thanks for the review and the kind words btw